### ISSUE TITLE
Avoid deserialising source of the get API key response when there is no change

### ISSUE THREAD

**Issue by ywangd** [elastic/elasticsearch#53940]
The get API key response is [deserialised](https://github.com/elastic/elasticsearch/blob/7.6/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java#L332) every time after the document is fetched. This could be a performance issue when the source document gets larger (e.g. many role descriptors or future derived API keys). One possible way to improve it is to check "primary term" and "sequence number" of the response and only deserialise the source if any of the two information have changed. Doing this also requires more information to be cached in the first read. So there is a balance between storage and speed.

**Comment by ywangd** [comment]
@tvernum Somehow this just occured to me, but why does the API key document needs to be fetched in the first place? Can we just actively clear the cache when a document is updated? Expiration date is a passive property that takes effect without a document update. But we could still cache the expiration date along with the key to avoid having the fetch the document?

**Comment by tvernum** [comment]
It doesn't but fetch-by-id is fast and distributed cache management is hard to get right so, on balance, the idea had been that always fetching the document is safer from a provable-correct point of view.

As you would have seen in the App Priv cache, you need to deal with new APIs to clear the cache and re-sync the cache if a node leaves the cluster temporarily, etc. etc.

The API Key service tries to avoid doing that by caching as little as possible and going back to the index each time. Given the issues you saw with the get threadpool, it might be worth re-thinking that tradeoff, but it requires getting the cache invalidation semantics perfectly correct.

**Comment by ywangd** [comment]
> re-sync the cache if a node leaves the cluster temporarily
>
Actually this is new to me. I haven't thought about this before. Thanks. I will add that for the Application privilege cache.

**Comment by tvernum** [comment]
> Actually this is new to me. I haven't thought about this before.

See 
https://github.com/elastic/elasticsearch/blob/c0ee68b0a03d0dafcd34bcb168f1e2d3d68f1886/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java#L453-L458

**Comment by ywangd** [comment]
Thanks Tim. Yes I added similar code and relevant tests for NativePrivilegeStore

https://github.com/elastic/elasticsearch/pull/55836/commits/07b7837c4284d40541d51eba66d431dc9680d80c

**Comment by ywangd** [comment]
**TLDR**: It would be helpful if **string representations instead of object are used for `roleDescriptors` and `limitedRoleDescriptors`** for the schema of API key document.

Skip parsing the API key document requires the document content to be cached so that it can be reused when the same key comes in again. There is obviously the balance between caching more with higher heap usage vs computational cost. An additional concern is when a large number of keys share the same set of role descriptors (which is a likely case for ingestion usage). In this situation, the repetitive caching of the almost identical document contents will likely offset any benefit from reducing role cache size (#53939).

Hence a two layer caching is needed, e.g.:
1. First layer mapping is: `ApiKeyId -> Tuple<ExpirationDate, UniqueIdofRoleDescriptors>`
2. Second layer mapping is: `UniqueIdOfRoleDescriptors -> RoleDescriptors of ApiKeyDocument`

where `UniqueIdofRoleDescriptors` guarantees that the same set of role descriptors always have the same ID regardless which documents they come frome. This can be achieved by hashing the content of `roleDescriptors` and `limitedRoleDescriptors`. Ideally, we do not want cache the entire API key document since the only relevant parts are `roleDescriptors` and `limitedRoleDescriptors`. However it is challenging to extract them out of a document. We could do it with string manipulation for `response.getSourceAsString()` (existing XContent utils do not help), but it could be fragile. If we, however, could afford changes to the schema of API key document so that string representations instead of object are used for `roleDescriptors` and `limitedRoleDescriptors`, it would help simplify the two layer caching implementation. 

Using string representations for `roleDescriptors` and `limitedRoleDescriptors` of an API key document can also help reducing existing de/serialisation cost. Currently both fields of an API key document are fully deserialised into `Map<String, Object>`. They then get serialised again into ThreadContext with `authenticationSerializer.writeToContext(...)`. If they are in string format, both the serialisation and deserialisation operations are no longer necessary. Also since the exact content of the descriptors are only of interest to `CompositeRolesStore`, it makes sense to defer the full desrialisation until Role building is needed.

**Comment by ywangd** [comment]
Another data point that I'd like to add is that GET API key documet takes about 60% of the time in the authentication process (*not* including authorization). Another thing to consider when we have the thread pool discussion, which currently does not have its own issue.

**Comment by ywangd** [comment]
`response.getSource()` and `authenticationSerializer.writeToContext(...)` _each_ take about 10% of the time in authentication process. Note these could very increase when the role descriptors get larger. These number are measure when `limitedRoleDescriptors` is from superuser and `roleDescriptors` is null.

**Comment by ywangd** [comment]
> TLDR: It would be helpful if string representations instead of object are used for roleDescriptors and limitedRoleDescriptors for the schema of API key document.
> 

Thanks @tvernum for prompting me. It is _not_ necessary. The existing XContent utils can partially deserialise an object (`ObjectParserHelper#declareRawObject`), i.e. it can leave an object field untouched and return the raw bytes. Therefore it is possible to achieve what I am proposing ([role cache keyed by hash of role descriptors' string representation](https://github.com/elastic/elasticsearch/issues/53939#issuecomment-643021514)) relatively easily without touching the document schema and additional de/serialisation cost.

**Comment by jaymode** [comment]
> Another thing to consider when we have the thread pool discussion, which currently does not have its own issue.

@ywangd can you elaborate on the threadpool issue/discussion or create an issue for it? I am asking because I am currently working on system indices having their own threadpool(s) for read operations and I'd like to understand if the work that I am doing could help here.

**Comment by ywangd** [comment]
@jaymode I created #58088 for it. Let me know if you need more information about it. Thanks
